bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttService.java

```diff
@@ -222,17 +231,17 @@ public boolean addBrokerConnection(MqttBrokerConnection connection) {
      * @throws ConfigurationException Most likely your provided name and url are invalid.
      * @throws MqttException
      */
-    public MqttBrokerConnection addBrokerConnection(Map<String, String> brokerConnectionConfig)
+    public @Nullable MqttBrokerConnection addBrokerConnection(Map<String, String> brokerConnectionConfig)
             throws ConfigurationException, MqttException {
         // Extract mandatory fields
         String brokerID = brokerConnectionConfig.get(NAME_PROPERTY);
-        if (StringUtils.isBlank(brokerID)) {
+        if (brokerID == null || brokerID.isEmpty()) {
             throw new ConfigurationException("MQTT Broker property 'name' is not provided");
         }
         brokerID = brokerID.toLowerCase();
 
         final String brokerURL = brokerConnectionConfig.get("url");
```

Comment:
another question about the non null behaviour of maps: With `@NonNullByDefault` we force the compiler to interpret a call to `brokerConnectionConfig.get("url")` to always return a non null value. How can this be guaranteed at all? I get the impression that any call to a map is interpreted as non null value, making additional null checks obsolete. Thats a dangerous behaviour:

```
Object value = myMap.get("key");
if (value == null) {
   throw ....
}
```

in this example the body of the if clause is reported as dead code. I guess for Maps the default should be `Map<@NoNull KEY_TYPE, @Nullable VALUE_TYPE>`.

wdyt?

Reply:
> another question about the non null behaviour of maps: With `@NonNullByDefault` we force the compiler to interpret a call to `brokerConnectionConfig.get("url")` to always return a non null value.

No, you define the default for your classes or packages but not for the map itself.
A map's get method could always return null e.g. your ask for a value of a non-existing key.

> I guess for Maps the default should be `Map<@NoNull KEY_TYPE, @Nullable VALUE_TYPE>`.

No, this is not correct. If you look at the Map API null is allowed for the key if the implementation allows null keys (stated in the JavaDoc).

Also `@NonNull K` is not correct at all, because of the upper / lower bounds of generic types. In the generic type declaration you need to use `K extends @NonNull Object`.

The checker framework has a good documentation about that. I realized it a few days ago.